### ISSUE TITLE
export PERLIO=bytes segfaults

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6548]
Migrated from [rt.perl.org#22390](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22390) (status was 'resolved')

Searchable as RT22390$

**Comment by p5pRT** [comment]
### From gp@familiehaase.de

#### Created by gp@familiehaase.de



I'm on Cygwin with latest release 1\.3\.22 and bleadperl&#8203;:

$ cat \.patch
19629

$ export PERLIO=bytes

$ perl \-e 'open OUTFILE\,">&#8203;:crlf"\,"text\.txt";print OUTFILE "\\n";'

this segfaults here\, perl should never segfault\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.9.0:

Configured by Gerrit at Fri May 30 20:08:16 EDT 2003.

Summary of my perl5 (revision 5.0 version 9 subversion 0 patch 19629) configuration:
  Platform:
    osname=cygwin, osvers=1.3.22(0.7832), archname=cygwin-thread-multi-64int
    uname='cygwin_nt-4.0 loreley 1.3.22(0.7832) 2003-03-18 09:20 i686 unknown unknown cygwin '
    config_args='-de -Dusedevel -Uversiononly -Dinstallusrbinperl -Dmksymlinks -Dusemultiplicity -Duse64bitint -Doptimize=-g -O3 -Dman3ext=3pm -Dusethreads -Uusemymalloc'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -DDEBUGGING -fno-strict-aliasing -I/usr/local/include',
    optimize='-g -O3',
    cppflags='-DPERL_USE_SAFE_PUTENV -DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.2.3 (home grown)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld2', ldflags =' -g -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -ldb -lcrypt -lutil
    perllibs=-lcrypt -lutil
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' -g'
    cccdlflags=' ', lddlflags=' -g -L/usr/local/lib'

Locally applied patches:
    DEVEL18374


@INC for perl v5.9.0:
    /usr/lib/perl5/5.9.0/cygwin-thread-multi-64int
    /usr/lib/perl5/5.9.0
    /usr/lib/perl5/site_perl/5.9.0/cygwin-thread-multi-64int
    /usr/lib/perl5/site_perl/5.9.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.9.0:
    CYGWIN=ntsec binmode tty nowinsymlinks error_start:h:\bin\gdb
    CYGWIN_ROOT=/
    HOME=/home/Gerrit
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=C
    LC_CTYPE=C
    LC_MESSAGES=C
    LD_LIBRARY_PATH=/usr/lib:/usr/local/lib:/usr/X11R6/lib
    LOGDIR=/home/logs
    PATH=/bin:/usr/bin:/usr/local/bin:/usr/X11R6/bin:/opt/gnome/bin:/sbin:/usr/sbin:/usr/local/sbin
    PERLIO=something_different
    PERL_BADLANG (unset)
    SHELL=/usr/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Gerrit P\. Haase \(via RT\) wrote&#8203;:
> 
> I'm on Cygwin with latest release 1\.3\.22 and bleadperl&#8203;:
> 
> $ cat \.patch
> 19629
> 
> $ export PERLIO=bytes
> 
> $ perl \-e 'open OUTFILE\,">&#8203;:crlf"\,"text\.txt";print OUTFILE "\\n";'
> 
> this segfaults here\, perl should never segfault\.

That's because the :bytes layer doesn't provide an "Open" method\.
\(Why ?\)

The naive patch below cures the segfault\, but it doesn't make your
open\(\) succeed\.

\-\-\- perlio\.c    \(revision 1426\)
\+\+\+ perlio\.c    \(working copy\)
@&#8203;@&#8203; \-3420\,8 \+3420\,10 @&#8203;@&#8203;
&nbsp;             \* mode\+\+;
&nbsp;             \*/
&nbsp;        \}
\-       f = \(\*tab\->Open\) \(aTHX\_ tab\, layers\, n \- 1\, mode\, fd\, imode\, perm\,
\-                         f\, narg\, args\);
\+       f = tab && tab\->Open
\+           ? \(\*tab\->Open\) \(aTHX\_ tab\, layers\, n \- 1\, mode\, fd\, imode\, perm\,
\+                         f\, narg\, args\)
\+           : Nullfp;
&nbsp;        if \(f\) \{
&nbsp;            if \(PerlIO\_push\(aTHX\_ f\, self\, mode\, PerlIOArg\) == 0\) \{
&nbsp;                /\*
End\.

**Comment by p5pRT** [comment]
### From @rgs

I wrote&#8203;:
> 
> That's because the :bytes layer doesn't provide an "Open" method\.
> \(Why ?\)

The docs answer my question : \(from perliol\.pod\)

&nbsp;    Having said all that translation\-only layers do not need to provide
&nbsp;    C\<Open\(\)> at all\, but rather leave the opening to a lower level layer
&nbsp;    and wait to be "pushed"\.  If a layer does provide C\<Open\(\)> it should
&nbsp;    normally call the C\<Open\(\)> method of next layer down \(if any\) and
&nbsp;    then push itself on top if that succeeds\.

Thus I propose to add a warning when doing an open\(\) with only translation
layers\. And to set $\!\. \(to what ? EINVAL ?\)

BTW the following appears to do what you want :

PERLIO=bytes perl \-e 'open OUTFILE\,">&#8203;:unix&#8203;:crlf"\,"text\.txt";print OUTFILE "\\n";'

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:
>Gerrit P\. Haase \(via RT\) wrote&#8203;:
>> 
>> I'm on Cygwin with latest release 1\.3\.22 and bleadperl&#8203;:
>> 
>> $ cat \.patch
>> 19629
>> 
>> $ export PERLIO=bytes
>> 
>> $ perl \-e 'open OUTFILE\,">&#8203;:crlf"\,"text\.txt";print OUTFILE "\\n";'
>> 
>> this segfaults here\, perl should never segfault\.
>
>That's because the :bytes layer doesn't provide an "Open" method\.
>\(Why ?\)

Because it is a pseudo layer that never ends up on the stack 
it just changes properties of layer\(s\) below\.

What is probably meant by the PERLIO= in the cygwin case 
is PERLIO="&#8203;:crlf&#8203;:bytes" or perhaps PERLIO="&#8203;:perlio&#8203;:bytes"

It should not segfault though\. I suppose if a PERLIO="" cannot open 
we should treat it as \_augmenting\_ the default stack rather then \_replacing\_\.


>
>The naive patch below cures the segfault\, but it doesn't make your
>open\(\) succeed\.
>
>\-\-\- perlio\.c    \(revision 1426\)
>\+\+\+ perlio\.c    \(working copy\)
>@&#8203;@&#8203; \-3420\,8 \+3420\,10 @&#8203;@&#8203;
>             \* mode\+\+;
>             \*/
>        \}
>\-       f = \(\*tab\->Open\) \(aTHX\_ tab\, layers\, n \- 1\, mode\, fd\, imode\, perm\,
>\-                         f\, narg\, args\);
>\+       f = tab && tab\->Open
>\+           ? \(\*tab\->Open\) \(aTHX\_ tab\, layers\, n \- 1\, mode\, fd\, imode\, perm\,
>\+                         f\, narg\, args\)
>\+           : Nullfp;
>        if \(f\) \{
>            if \(PerlIO\_push\(aTHX\_ f\, self\, mode\, PerlIOArg\) == 0\) \{
>                /\*
>End\.
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @rgs

Jarkko cured the segfault as change \#19747\.

But\, as he says\, "TODO&#8203;: we really should not get here if the layer stack
is corrupt or non\-sensical \(like "bytes"\)\."

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @jhi

I forgot to mention that I patched this in change \#19747\, using more
or less Rafael's testing for \->Open\, amended by NI\-S' private suggestion
of setting the errno\.  This is however just a safety net&#8203;: logically speaking
we shouldn't get this far in the code if the PERLIO is illegal\, like "bytes"
or "sesquipedalian"\.

\-\- 
Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> http&#8203;://www\.iki\.fi/jhi/ "There is this special
biologist word we use for 'stable'\.  It is 'dead'\." \-\- Jack Cohen

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
